server/src/main/java/org/apache/iotdb/db/mpp/buffer/SourceHandle.java

```diff
@@ -19,72 +19,286 @@
+      DataBlockService.Client client,
+      TsBlockSerde serde,
+      SourceHandleListener sourceHandleListener) {
+    this.remoteHostname = Validate.notNull(remoteHostname);
+    this.remoteFragmentInstanceId = Validate.notNull(remoteFragmentInstanceId);
+    this.localFragmentInstanceId = Validate.notNull(localFragmentInstanceId);
+    this.localOperatorId = Validate.notNull(localOperatorId);
+    this.localMemoryManager = Validate.notNull(localMemoryManager);
+    this.executorService = Validate.notNull(executorService);
+    this.client = Validate.notNull(client);
+    this.serde = Validate.notNull(serde);
+    this.sourceHandleListener = Validate.notNull(sourceHandleListener);
+    bufferRetainedSizeInBytes = 0L;
   }
 
   @Override
-  public TsBlock receive() {
+  public synchronized TsBlock receive() throws IOException {
     if (throwable != null) {
-      throw new RuntimeException(throwable);
+      throw new IOException(throwable);
     }
     if (closed) {
-      throw new IllegalStateException("Source handle has been closed.");
+      throw new IllegalStateException("Source handle is closed.");
+    }
+    if (!blocked.isDone()) {
+      throw new IllegalStateException("Source handle is blocked.");
     }
     TsBlock tsBlock = bufferedTsBlocks.poll();
     if (tsBlock != null) {
-      bufferRetainedSizeInBytes -= getRetainedSizeInBytes(tsBlock);
+      bufferRetainedSizeInBytes -= tsBlock.getRetainedSizeInBytes();
+      localMemoryManager
+          .getQueryPool()
+          .free(localFragmentInstanceId.getQueryId(), tsBlock.getRetainedSizeInBytes());
     }
-    if (bufferedTsBlocks.isEmpty() && !finished && blocked.isDone()) {
+    if (bufferedTsBlocks.isEmpty() && !isFinished() && blocked.isDone()) {
       blocked = SettableFuture.create();
     }
+    if (isFinished()) {
+      sourceHandleListener.onFinished(this);
+    }
+    trySubmitGetDataBlocksTask();
     return tsBlock;
   }
 
-  private long getRetainedSizeInBytes(TsBlock tsBlock) {
-    throw new UnsupportedOperationException();
+  private void trySubmitGetDataBlocksTask() {
+    if (getDataBlocksTaskFuture != null && !getDataBlocksTaskFuture.isDone()) {
+      return;
+    }
+    int startSequenceId = nextSequenceId;
+    int currSequenceId = nextSequenceId;
+    long reservedBytes = 0L;
+    while (sequenceIdToDataBlockSize.containsKey(currSequenceId)) {
+      Long bytesToReserve = sequenceIdToDataBlockSize.get(currSequenceId);
+      if (bytesToReserve == null) {
+        throw new IllegalStateException("Data block size is null.");
+      }
+      boolean reserved =
+          localMemoryManager
+              .getQueryPool()
+              .tryReserve(localFragmentInstanceId.getQueryId(), bytesToReserve);
+      if (reserved) {
+        currSequenceId += 1;
+        reservedBytes += bytesToReserve;
+        bufferRetainedSizeInBytes += bytesToReserve;
+      } else {
+        break;
+      }
+    }
+
+    if (currSequenceId > startSequenceId) {
```

Comment:
If we cannot get enough memory in one try, it seems that there is no more event to notify the SourceHandle to try again if the memory is released

Reply:
trySubmitGetDataBlocksTask will be called in 2 cases:

1. new data blocks generated by upstream instance -> see SourceHandle#updatePendingDataBlockInfo
2. memory released -> see SourceHandle#receive

Reply:
Fixed.